Expose torch.empty(sizes, *, names, ...) to Python

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21795 Implement tensor.select(Dimname,int)
* #21789 Add at::dimname_to_position helper.
* #21784 Move at::internal_set_names_inplace to ATen/NamedTensorUtils.h
* #21752 named inference rule for tensor.select
* #21781 Disallow creation of tensors with duplicate names
* #21735 Copy NamedTensorMeta in TensorImpl::copy_tensor_data()
* **#21648 Expose torch.empty(sizes, *, names, ...) to Python**

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

Differential Revision: [D15804482](https://our.internmc.facebook.com/intern/diff/D15804482)

zou3519 committed:
Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

nairbv commented on torch/csrc/python_dimname.cpp:
```diff
@@ -0,0 +1,83 @@
+    return *maybeDimname;
+  }
+
+  const auto name = THPUtils_unpackString(obj);
```
I'm not sure if I'm understanding this correctly... first we ensure the python string is interned, but then if we don't already have a dimname in python this unpack will create a new uninterned C++ string and intern that too right? 

In the implementation of `THUtils_unpackString` I see `return std::string(PyBytes_AS_STRING(obj), size);`

Do we have two full separate copies of tables of interned strings and their dim names, one in python and one in c++? Is there any way to share interned string representations between python/c++?

zou3519 replied:
> I'm not sure if I'm understanding this correctly... first we ensure the python string is interned, but then if we don't already have a dimname in python this unpack will create a new uninterned C++ string and intern that too right?

Yes.

> Do we have two full separate copies of tables of interned strings and their dim names, one in python and one in c++? Is there any way to share interned string representations between python/c++?

In C++ we have Symbols, which are interned strings in C++. In Python we have "python interned strings". Right now the table the PythonArgParser holds is used to translate between Python interned strings and C++ Symbols.

Python interned strings and Symbols are different. I don't know of a good way to unify them (also because Symbol has to be Python-agnostic since we compile a libtorch standalone...).

nairbv replied:
hmm.. are they sharing any memory? it seems like `PyBytesObject` and `std::string` are both wrapping a `char*`.

zou3519 replied:
We can't pass around `char*` in place of `Dimname` because we need the full `Dimname` struct.

Replacing all instances of `Symbol` (in Dimname) with `char*` is a little complicated. It's hard to guarantee that a string passed in from python is the same as a c++ interned string.

nairbv replied:
What I meant was, I'm just curious if we actually have two of the same char*'s in memory, or if the dimname->symbol->string->char* and pyobject->char* both reference the same memory

zou3519 replied:
> if the dimname->symbol->string->char* and pyobject->char* both reference the same memory

These could be different char*, or they could be the same char*. They're the same char* if the user only interacts with PyTorch via python and the name has not already been used in cpp-land.

If the name has been used in cpp-land (but not python land), and then the user creates a new name in python land, then those two names have different char*.

nairbv replied:
ah that makes sense. thanks

nairbv commented on torch/csrc/python_dimname.cpp:
```diff
@@ -0,0 +1,83 @@
+    // See https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_InternInPlace
+    Py_INCREF(obj);
+    THPUtils_internStringInPlace(&obj);
+    Py_DECREF(obj);
```
is there somehow we can test this interning system / that it properly allows freeing the memory for the dim names etc? I'd imagine it's not a very common case that a dim name goes completely out of scope and stops being used, and dim names are small, so would be easy to not notice if this doesn't actually work correctly.

zou3519 replied:
I don't have a good idea for how to test this. It's not easy to use mocks because we have global state running around in a number of places. I'll go and think about this a bit more but I am taking suggestions

zou3519 replied:
Okay, might be possible to test this in Python: https://stackoverflow.com/questions/510406/is-there-a-way-to-get-the-current-ref-count-of-an-object-in-python . Going to see what I can do here.

nairbv replied:
It looks like we can count total reference counts using sys.gettotalrefcount()

https://pythonextensionpatterns.readthedocs.io/en/latest/debugging/leak_newrefs_vg.html#observing-the-total-reference-counts

We could verify it goes up/down by an expected amount if we create random dimname strings and then let them go out of scope.

zou3519 replied:
Yup, I added some tests for the refcounting behavior.

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head

zou3519 committed:
Update on "Expose torch.empty(sizes, *, names, ...) to Python"

Expose torch.empty(sizes, *, names, ...) to Python

There are a few main components to this:

Implement `tensor.names` in Python.
- This results a `List[Optional[str]]`. In the future it'll return a
`List[Optional[torch.Dimname]]` when we implement torch.Dimname objects,
but now it is like this for testing.

PythonArgParser learns how to handle `DimnameList?`
- PythonArgParser stores an InternedStringsTable that maps interned
PyObject* to at::Dimname objects. In cpp-land we use at::Dimname, in
python land we work with strings.
- PythonArgParser converts input Nones and strings to at::Dimname.
- None gets converted to Dimname::wildcard()
- PythonArgParser interns a python string if it is not interned (the
mapping only stores interned strings).
- If the interned string is not in the table, PythonArgParser parses it
to get an at::Dimname and adds the (interned string, Dimname) mapping to
the table.

![image](https://user-images.githubusercontent.com/5652049/59299353-7175c680-8c5a-11e9-8f57-220cce9cbec7.png)

unpacking `optional<DimnameList>`.
- We get an `optional<vector<Dimname>>` from `PythonArgs`. However, this
needs to be converted to an `optional<ArrayRef<Dimname>>` (because that
is what `at::empty` takes as an argument).
- To do this, I added extra codegen to unwrap
`optional<vector<Dimname>>` to `vector<Dimname>>`, convert that to
`ArrayRef<Dimname>`, and then rewrap it to
`optional<ArrayRef<Dimname>>`.

Test Plan:
- `python test/test_namedtensor.py -v`
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21648 gh/zou3519/48/head
